### ISSUE TITLE
Fundamental error in the code

### ISSUE THREAD

**Issue by HugoFry** [mashaan14/VisionTransformer-MNIST#1]
The following line of code in notebook I believe is incorrect:

`transformer_input_expanded = model.transformer[0].linear[0](transformer_input)[0]`

This is taking the hidden state of the MLP ('linear' module), not the attention activations. This means the qkv are formed from the MLP activations not MHA. The qk values are also not softmaxed, which is necessary to obtain the contributions of each token (softmax is invariant under addition of a constant).

**Comment by mashaan14** [comment]
Thanks for your comment,

I'm pulling the values at the MLP block inside the transformer encoder.

<img width="236" alt="Screenshot 2024-07-18 at 2 12 33 PM" src="https://github.com/user-attachments/assets/f10a0309-019c-4b87-9476-9f4e0663d7a9">

In the code, there're 6 layers of transformer encoder, and I'm pulling these values at the first layer.

**Comment by HugoFry** [comment]
Thanks for the quick response!

If my understanding serves me correctly, your repository is attempting to performing attention rollout on the ViT is that correct? If so, you want to obtain the attention patterns from the attention heads. Your code is hooking the activations of the MLP, **NOT** the attention heads. Your code also defines query/key/value variables in terms of the MLP activations, but these are just pretty random pre-GELU MLP activations and as such have nothing to do with the ture QKV computations. This means the patterns you visualise have nothing to do with where attention heads are moving information and are therefore kind of meaningless. You also haven't correctly hooked the model since the layer 0 MLP is not given the correct activations (you've neglected the layer norm **AND** the layer 0 attention heads). I could well have misunderstood what you're trying to do here though.

Let me know if I've misunderstood. If I'm correct in pointing out these errors, I'm happy to make a pull request with the corrections if you'd like and then you can review my changes?

**Comment by mashaan14** [comment]
The question I'm trying to answer is:
Given an unseen image (picked randomly from the test set), how the trained model would react?

Of course you could inspect each head individually. But I want to see by the end of the first layer, how did it react to the presented test image?

The code is in a jupyter notebook, I'm not sure if you can edit that, but you're welcome to submit a pull request.
